File: #82_Remove_Duplicates_from_Sorted_List_II/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) return head;
        ListNode dummyHead = new ListNode(0);
        dummyHead.next = head;
        ListNode prev = dummyHead;
        ListNode cur = head;
        while (cur != null) {
            // duplicate
            while (cur.next != null && cur.next.val == cur.val) {
                cur = cur.next;
            }
            if (prev.next == cur) {
                prev = cur;
            }
            else {
                prev.next = cur.next;
            }
            cur = cur.next;
        }
        return dummyHead.next;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	ListNode head = new ListNode(1);
    	ListNode cur = head;
    	cur.next = new ListNode(1);
    	cur = cur.next;
    	cur.next = new ListNode(1);
    	cur = cur.next;
    	cur.next = new ListNode(1);
    	cur = cur.next;
    	ListNode newHead = sol.deleteDuplicates(head);
    	cur = newHead;
    	while(cur != null) {
    		System.out.println(cur.val);
    		cur = cur.next;
    	}
    }
}

File: #82_Remove_Duplicates_from_Sorted_List_II/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def deleteDuplicates(self, head):
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        cur = head
        dup = False
        while cur and cur.next:
        	if cur.next.val != cur.val:
        		cur = cur.next
        		prev = prev.next
        	else:
        		dup = True
        		while cur.next and cur.next.val == cur.val:
        			cur = cur.next
        		prev.next = cur.next
        		cur.next = None
        		cur = prev.next
        		dup = False
        return dummy.next

sol = Solution()
h = ListNode(1)
n1 = ListNode(1)
n2 = ListNode(2)
n3 = ListNode(3)
n4 = ListNode(3)
h.next = n1
n1.next = n2
n2.next = n3
n3.next = n4
cur = sol.deleteDuplicates(h)
while cur:
	print cur.val
	cur = cur.next